### PR TITLE
hr_employee_id port to v8

### PR DESCRIPTION
**Pull request by saltonmassally**
Initial Attempt of a port to V8
Some improvements made are as follows:
- Ability to choose ID generation method
- Sequence vs Random

### CODE REVIEW

File: hr_employee_id/README.rst [hunk 1]
```diff
@@ -0,0 +1,69 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
+
+Employee Identification Numbers
```
Comment (pedrobaeza) [inline hr_employee_id/README.rst:6]: Put also top level line

File: hr_employee_id/__init__.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# -*- coding:utf-8 -*-
+#
+#
+#    Copyright (C) 2011,2013 Michael Telahun Makonnen <mmakonnen@gmail.com>.
+#    All Rights Reserved.
```
Comment (pedrobaeza) [inline hr_employee_id/__init__.py]: Remove this line

File: hr_employee_id/__openerp__.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
+# -*- coding:utf-8 -*-
+#
+#
+#    Copyright (C) 2011,2013 Michael Telahun Makonnen <mmakonnen@gmail.com>.
+#    All Rights Reserved.
```
Comment (pedrobaeza) [inline hr_employee_id/__openerp__.py]: Remove this line

File: hr_employee_id/__openerp__.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#
+
+{
+    'name': 'Employee ID',
+    'version': '1.0',
+    'category': 'Generic Modules/Human Resources',
+    'description': """
```
Comment (pedrobaeza) [inline hr_employee_id/__openerp__.py]: Remove description section

File: hr_employee_id/__openerp__.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+Employee Identification Numbers
+===============================
+Company wide unique employee ID
+    """,
+    'author': "Michael Telahun Makonnen <mmakonnen@gmail.com>, "
+    "Odoo Community Association (OCA), "
```
Comment (pedrobaeza) [inline hr_employee_id/__openerp__.py]: Indent this at the same level at the beginning of the string

File: hr_employee_id/models/__init__.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
+# -*- coding:utf-8 -*-
+##############################################################################
+#
+#    Copyright (C) 2011,2013 Michael Telahun Makonnen <mmakonnen@gmail.com>.
+#    All Rights Reserved.
```
Comment (pedrobaeza) [inline hr_employee_id/models/__init__.py]: Remove this line

File: hr_employee_id/models/hr_employee.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    
+    _inherit = 'hr.employee'  
+    
+
+    employee_no = fields.Char('Employee ID', required=False,
+                       readonly=True, copy=False, default='/')
+
+
+    _sql_constraints = [
+        ('employeeno_uniq', 'unique(employee_no)', 'The Employee Number must be unique across the company(s).'),
```
Comment (pedrobaeza) [inline hr_employee_id/models/hr_employee.py]: Please respect 80 cols length limit

File: hr_employee_id/models/res_config.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    employee_id_gen_method = fields.Selection((('random', 'Random'),
+                                               ('sequence', 'Sequence')), string="ID Generation Method", default='random',
+                                              help="Method by which employee ID is generated; \
+                                              Random - IDs are generated randomly \
+                                              Sequence - IDs are generated based on a supplied sequence")
+    
+    employee_id_random_digits = fields.Integer('# of Digits', default=5, help="Number of digits making up the ID")
+    employee_id_sequence = fields.Many2one('ir.sequence', 'Sequence', help="Pattern to be used for used for ID Generation")
+    
+    def get_default_employee_id_gen_method(self, cr, uid, ids, context=None):
```
Comment (pedrobaeza) [inline hr_employee_id/models/res_config.py]: Use new API defaults

File: hr_employee_id/README.rst (hunk 1 above)
Comment (saltonmassally) [inline hr_employee_id/README.rst:6]: @pedrobaeza thank you for taking the time to look at this; I don't understand what you mean by this; looked through other READMEs in OCA/hr and they look similar...

P.S. THis is my first OCA contrib so forgive me I am unaware of these seemly apparent conventions

Comment (pedrobaeza) [inline hr_employee_id/README.rst:6]: OK, don't worry. We are here to help. I mean to copy the same line `=====` at the top. This is a very recent addition, because we were not doing well the styles. Look at this PR to understand it: https://github.com/OCA/maintainer-tools/pull/109

**Comment by saltonmassally** [comment]
Sorry I have been away from this but back to it.
At @pedrobaeza do you see the need for employee_no? identification_id is already provided by hr addon  do you reckon that we should use this instead?

File: hr_employee_id/models/hr_employee.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+from openerp import models, fields, api, _
+from openerp.exceptions import Warning
+
+
+class HrEmployee(models.Model):
+    """Implement company wide unique identification number."""
+
+    _inherit = 'hr.employee'
+
+    employee_no = fields.Char('Employee ID', required=False,
```
Comment (ehdem) [inline hr_employee_id/models/hr_employee.py]: @tarzan0820 : No need required=False

File: hr_employee_id/models/hr_employee.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+            tries = 0
+            max_tries = 50
+            while tries < max_tries:
+                rnd = random.SystemRandom()
+                employee_id = ''.join(rnd.choice(string.digits)
+                                      for _ in xrange(employee_id_random_digits))
+                self.env.cr.execute(
+                    '''SELECT employee_no FROM hr_employee WHERE
+                     employee_no=%s''', tuple((employee_id,)))
+                res = self.env.cr.fetchall()
```
Comment (ehdem) [inline hr_employee_id/models/hr_employee.py]: @tarzan0820 : Why you don't use ORM ?

File: hr_employee_id/models/res_config.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+##############################################################################
+
+from openerp import models, fields, SUPERUSER_ID
+
+
+class human_resources_configuration(models.TransientModel):
+    _inherit = 'hr.config.settings'
+
+    employee_id_gen_method = fields.Selection((('random', 'Random'),
+                                               ('sequence', 'Sequence')), string="ID Generation Method", default='random',
```
Comment (ehdem) [inline hr_employee_id/models/res_config.py]: @tarzan0820 : use this: fields.Selection(
        [
            ('random', 'Random'),
            ('sequence', 'Sequence'),
        ],
.....

File: hr_employee_id/models/res_config.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+                                              Sequence - IDs are generated based on a supplied sequence")
+    
+    employee_id_random_digits = fields.Integer('# of Digits', default=5, help="Number of digits making up the ID")
+    employee_id_sequence = fields.Many2one('ir.sequence', 'Sequence', help="Pattern to be used for used for ID Generation")
+    
+    def get_default_employee_id_gen_method(self, cr, uid, ids, context=None):
+        employee_id_gen_method  = self.pool.get("ir.config_parameter").get_param(cr, uid, "hr_employee_id.employee_id_gen_method", default=None, context=context)
+        return {'employee_id_gen_method': employee_id_gen_method  or False}
+
+    def set_employee_id_gen_method (self, cr, uid, ids, context=None):
```
Comment (ehdem) [inline hr_employee_id/models/res_config.py]: @tarzan0820 : Use new API

File: hr_employee_id/models/res_config.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+        config_parameters = self.pool.get("ir.config_parameter")
+        for record in self.browse(cr, uid, ids, context=context):
+            config_parameters.set_param(cr, uid, "hr_employee_id.employee_id_gen_method", record.employee_id_gen_method  or '', context=context)
+            
+    def get_default_employee_id_random_digits(self, cr, uid, ids, context=None):
```
Comment (ehdem) [inline hr_employee_id/models/res_config.py]: @tarzan0820 : Use new API

File: hr_employee_id/models/res_config.py [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+        employee_id_random_digits  = self.pool.get("ir.config_parameter").get_param(cr, uid, "hr_employee_id.employee_id_random_digits", default=None, context=context)
+        return {'employee_id_random_digits': int(employee_id_random_digits)  or False}
+
+    def set_employee_id_random_digits (self, cr, uid, ids, context=None):
```
Comment (ehdem) [inline hr_employee_id/models/res_config.py]: @tarzan0820 : Use new API

File: hr_employee_id/models/res_config.py [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+        config_parameters = self.pool.get("ir.config_parameter")
+        for record in self.browse(cr, uid, ids, context=context):
+            config_parameters.set_param(cr, uid, "hr_employee_id.employee_id_random_digits", record.employee_id_random_digits  or '', context=context)
+            
+    def get_default_employee_id_sequence(self, cr, uid, ids, context=None):
```
Comment (ehdem) [inline hr_employee_id/models/res_config.py]: @tarzan0820 : Use new API

File: hr_employee_id/models/res_config.py [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+        employee_id_sequence  = self.pool.get("ir.config_parameter").get_param(cr, uid, "hr_employee_id.employee_id_sequence", default=None, context=context)
+        if not employee_id_sequence:
+            dataobj = self.pool.get('ir.model.data')
+            try:
+                dummy, employee_id_sequence = dataobj.get_object_reference(cr, SUPERUSER_ID, 'hr_employee_id', 'seq_employeeid_ref')
+            except ValueError:
+                pass
+        return {'employee_id_sequence': int(employee_id_sequence)  or False}
+
+    def set_employee_id_sequence (self, cr, uid, ids, context=None):
```
Comment (ehdem) [inline hr_employee_id/models/res_config.py]: @tarzan0820 : Use new API

**Comment by ehdem** [comment]
@tarzan0820 : Fix travis errors

**Comment by dreispt** [comment]
Needs a rebase.

**Comment by saltonmassally** [comment]
@dreispt rebased

**Comment by dreispt** [comment]
@moylop260 @pedrobaeza Something's wrong here: lint is failing because of new checks, but AFAIK those are in PR still to be merged. I'm confused, can you help?

**Comment by moylop260** [comment]
Duplicated key in manifest dont is a beta message.

**Comment by saltonmassally** [comment]
@dreispt Travis is green

**Comment by moylop260** [comment]
@tarzan0820 
Could you check https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#idioms?
`Exceptions: Use from openerp.exceptions import Warning as UserError (v8) or from openerp.exceptions import UserError (v9) or find a more appropriate exception in openerp.exceptions.py`

``` bash
hr_employee_id/models/hr_employee.py:23: [R8101(openerp-exception-warning), ] Import `Warning` should be renamed as UserError `from openerp.exceptions import Warning as UserError`
```

Because `Warning` is a built-in function of python

**Comment by andhit-r** [comment]
@tarzan0820 any progress on this. May i offer help to finish it?

**Comment by saltonmassally** [comment]
@andhit-r I can push my latest changes and you can finish it off.... will do it in a few minutes

**Comment by andhit-r** [comment]
Great. I will continue your work.

**Comment by andhit-r** [comment]
@tarzan0820 I create PR on your branch https://github.com/tarzan0820/hr/pull/1

**Comment by saltonmassally** [comment]
@andhit-r merged

**Comment by andhit-r** [comment]
Thanks @tarzan0820.

@pedrobaeza @moylop260 @dreispt this PR is ready to be reviewed.

**Comment by dreispt** [comment]
Hi @tarzan0820, I just noticed that the changeset is in incorrect.
I think you need to rebase on top of latest 8.0 branch.

**Comment by andhit-r** [comment]
@dreispt @tarzan0820 I will rebase it

**Comment by andhit-r** [comment]
@tarzan0820 @dreispt  I rebased from upstream on https://github.com/tarzan0820/hr/pull/2

**Comment by andhit-r** [comment]
@dreispt @pedrobaeza @moylop260 rebased and ready to be reviewed

**Comment by andhit-r** [comment]
@OCA/human-resources-maintainers would someone kindly review this PR?

File: hr_contract_reference/views/hr_contract_view.xml [hunk 17]
```diff
@@ -1,11 +1,10 @@
 <?xml version="1.0" encoding="utf-8"?>
 <openerp>
     <data>
-        
+
```
Comment (pedrobaeza) [inline hr_contract_reference/views/hr_contract_view.xml:4]: Why these changes in this file?

File: hr_employee_id/README.rst [hunk 18]
```diff
@@ -0,0 +1,74 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+============================
+Employee Identification Numbers
+============================
```
Comment (pedrobaeza) [inline hr_employee_id/README.rst]: Put the same length of the title

File: hr_employee_id/README.rst [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+
+Company wide unique employee ID. Supports
+* Random ID Generation
```
Comment (pedrobaeza) [inline hr_employee_id/README.rst:11]: Add a line break here for applying the style

File: hr_employee_id/models/hr_employee.py [hunk 20]
```diff
@@ -0,0 +1,58 @@
+# -*- coding: utf-8 -*-
+# © 2011, 2013 Michael Telahun Makonnen <mmakonnen@gmail.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+import random
+import string
```
Comment (pedrobaeza) [inline hr_employee_id/models/hr_employee.py:6]: Why this import?

File: hr_employee_id/views/hr_employee_views.xml [hunk 21]
```diff
@@ -0,0 +1,58 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<openerp>
+	<data>
+
+		<record id="view_employee_form" model="ir.ui.view">
```
Comment (pedrobaeza) [inline hr_employee_id/views/hr_employee_views.xml]: Don't use tabs, but spaces in XMLs

File: hr_employee_id/models/hr_employee.py (hunk 20 above)
Comment (andhit-r) [inline hr_employee_id/models/hr_employee.py:6]: Line 40. **string.digits** = '0123456789'

File: hr_contract_reference/views/hr_contract_view.xml (hunk 17 above)
Comment (andhit-r) [inline hr_contract_reference/views/hr_contract_view.xml:4]: I must have missed it when rebase & resolving conflict. Is it ok?

Comment (pedrobaeza) [inline hr_contract_reference/views/hr_contract_view.xml:4]: OK, there's no special problem with it indeed.

File: hr_employee_id/models/hr_employee.py (hunk 20 above)
Comment (pedrobaeza) [inline hr_employee_id/models/hr_employee.py:6]: OK

**Comment by andhit-r** [comment]
@pedrobaeza modify and ready to be reviewed again

**Comment by pedrobaeza** [comment]
:+1:, but why there are 49 commits. Isn't too much for this simple module?

**Comment by andhit-r** [comment]
@pedrobaeza i looked into this PR's commit history and found out that it has messy history due rebase process. I tried to cherry pick commit @tarzan0820 and my work on this module. The result is https://github.com/andhit-r/hr/commits/8.0-hr_employee_id_cherry_pick. Should we use that PR instead of this?

**Comment by pedrobaeza** [comment]
Yes, maybe better, but you can also squash some of them in one from your part that makes little changes.

**Comment by andhit-r** [comment]
Hi @pedrobaeza thanks for your input. Would you kindly take a look to this https://github.com/andhit-r/hr/tree/8.0-hr_employee_id_cp_squash. I squash my work.

If it's okay i will create PR from that branch. 

Thanks in advance

**Comment by pedrobaeza** [comment]
Thanks for the efforts, @andhit-r. It looks goods, and it will look better squashing also some of the commits of @tarzan0820, but I'm asking too much, so if you want to give up now, it's alright.

**Comment by andhit-r** [comment]
No worries @pedrobaeza . Squash into 2 commits, @tarzan0820 and my work. Would you kindly review it again https://github.com/andhit-r/hr/tree/8.0-hr_employee_id_4

**Comment by pedrobaeza** [comment]
Perfect! We can close this one and propose the other then.

**Comment by andhit-r** [comment]
Ok i created new PR ( #205 ). Thanks for your help @pedrobaeza.
